fact/time.py

```diff
@@ -30,112 +20,126 @@
+
+    Parameters
+    ----------
+    dt: int, float, np.ndarray, pd.Series
+        The datetime object to convert to mjd
+    epoch: datetime
+        The epoch, default is classic MJD (1858-11-17T00:00)
+    '''
+    if isinstance(jd, (int, float)):
+        delta = timedelta(microseconds=jd * 24 * 3600 * 1e6)
+        return epoch + delta
 
-    if isinstance(mjd, pd.Series):
-        delta = (mjd * 24 * 3600 * 1e9).astype('timedelta64[ns]')
-        return MJD_EPOCH + delta
+    if isinstance(jd, pd.Series):
+        delta = (jd * 24 * 3600 * 1e9).astype('timedelta64[ns]')
+        return epoch + delta
 
     # other types will be returned as numpy array if possible
-    mjd = np.asanyarray(mjd)
-    delta = (mjd * 24 * 3600 * 1e9).astype('timedelta64[ns]')
-    return MJD_EPOCH_NUMPY + delta
+    jd = np.asanyarray(jd)
+    delta = (jd * 24 * 3600 * 1e9).astype('timedelta64[ns]')
+    return datetime_to_numpy(epoch) + delta
 
 
-def fjd(datetime_inst):
-    ''' convert datetime instance to FJD
+def fjd_to_datetime(fjd):
     '''
-    if datetime_inst.tzinfo is None:
-        logging.warning('datetime instance is not aware of its timezone.'
-                        'result possibly wrong!')
-    return calendar.timegm(datetime_inst.utctimetuple()) / (24. * 3600.)
-
-
-def iso2dt(iso_time_string):
-    ''' parse ISO time string to timezone aware datetime instance
-
-    example
-    2015-01-23T08:08+01:00
+    Convert a FACT julian date float to datetime, epoch is 1970-01-01T00:00Z
+    This function can handle python int, float, numpy arrays and pandas Series.
     '''
-    datetime_inst = dateutil.parser.parse(iso_time_string)
-    # make aware at any cost!
-    if datetime_inst.tzinfo is None:
-        logging.info('ISO time string did not contain timezone info. I assume UTC!')
-        datetime_inst = datetime_inst.replace(tzinfo=dateutil.tz.tzutc())
-    return datetime_inst
+    return mjd_to_datetime(fjd, epoch=UNIX_EPOCH)
 
 
-def run2dt(run_string):
-    ''' parse typical FACT run file path string to datetime instance (UTC)
-
-    example
-    first you do this:
-
-    '/path/to/file/20141231.more_text' --> '20141231'
-    then call
-    run2dt('20141231')
+def datetime_to_fjd(dt):
     '''
-    format_ = '%Y%m%d'
-    datetime_inst = datetime.strptime(run_string, format_)
-    datetime_inst = datetime_inst.replace(tzinfo=dateutil.tz.tzutc())
-    return datetime_inst
-
-
-def facttime(time_string):
-    ''' conver time-string to fact time
+    Convert a datetime to FACT julian date float, epoch is 1970-01-01T00:00Z.
+    This function can handle python dates, numpy arrays and pandas Series.
     '''
-    return calendar.timegm(time.strptime(
-        time_string, '%Y%m%d %H:%M')) / (24. * 3600.)
+    return datetime_to_mjd(dt, epoch=UNIX_EPOCH)
 
 
-def to_datetime(fact_julian_date):
-    ''' convert facttime to datetime instance
+def iso_to_datetime(iso):
+    '''
```

Comment:
there is a python package called `iso8601` ..  could we use this here?

Reply:
https://pypi.org/project/iso8601/

Reply:
Why? All the parsing is done by pandas, this just checks if the input can be read by pandas and transforms it if necessary.

Reply:
I thought this: We use `dateutil.parser.parse` which is pretty relaxed in what it parses:
```
str_us_format = "11/01/2012 17:21:00"
str_de_format = "11.01.2012 17:21:00"

assert dateutil.parser.parse(str_us_format) == dateutil.parser.parse(str_de_format)
# does not raise .. both are parsed into the same datetime instance
# even though humans might have meant two different dates.
# this is: mistakes waiting to happen.
```
While `iso8601` is more strict:
```
iso8601.parse_date(str_de_format)  # --> raises ParseError

str_iso_format = '2012-01-11 17:21:00'
iso8601.parse_date(str_iso_format)  # works
#Result: datetime.datetime(2012, 1, 19, 17, 21, tzinfo=datetime.timezone.utc)
```
Also note: `iso8601` sets the timezone to utc, if it is not in the string, which is the same behaviour we also added manually. I am not sure if this is exactly correct according to the Wiki ISO8601 article, but I did not study it in detail.

Reply:
Ah I just see, `pd.to_datetime([str_de_format, str_us_format, str_iso_format])` creates the same time for each of these strings .. so it is also pretty relaxed in its parsing...

So using `iso8601` would not create a consistent experience... 

Hm .. where do the strings come from? Fits headers, right?

Reply:
I would say arbitrary, as this is a top level function in a general purpose library. 

I think in pyfact it's only used in the io module reading from h5py files

Reply:
I do not really understand what you said there, bit since iso8601 would destroy a consistent user experience. I agree that it is not good to use it here.